src/pkg/signing/signing.go

```diff
@@ -78,25 +64,49 @@ func sifAddSignature(fingerprint [20]byte, fimg *sif.FileImage, signature []byte
 // configuration options. In its current form, Sign also pushes public material
 // to a key server if enabled. This should be a separate step in the next round
 // of development.
-func Sign(cpath, authToken string) error {
-	var el openpgp.EntityList
-	var en *openpgp.Entity
-	var err error
-
-	if el, err = sypgp.LoadPrivKeyring(); err != nil {
+func Sign(cpath, url, authToken string) error {
+	elist, err := sypgp.LoadPrivKeyring()
+	if err != nil {
 		return err
-	} else if el == nil {
-		return fmt.Errorf("no private keys found in %s, run 'singularity keys newpair' to create keys", sypgp.SecretPath())
 	}
 
-	if len(el) > 1 {
-		if en, err = sypgp.SelectPrivKey(el); err != nil {
+	var entity *openpgp.Entity
+	if elist == nil {
+		resp, err := sypgp.AskQuestion("No OpenPGP signing keys found, autogenerate? [Y/n] ")
+		if err != nil {
+			return err
+		}
+		if resp == "" || resp == "y" || resp == "Y" {
+			entity, err = sypgp.GenKeyPair()
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("cannot sign without installed keys")
+		}
+		resp, err = sypgp.AskQuestion("Upload public key %X to key server? [Y/n] ", entity.PrimaryKey.Fingerprint)
```

Comment:
It would be nice to see the URL that will be used in this message.

Reply:
Great now it's wrapping to the next output line ;-) Done.